core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ParsedHistogram.java

```diff
@@ -0,0 +1,151 @@
+                }
+            }
+
+        }, CommonFields.BUCKETS, ObjectParser.ValueType.OBJECT_ARRAY);
+    }
+
+    public static ParsedHistogram fromXContent(XContentParser parser, String name) throws IOException {
+        ParsedHistogram aggregation = PARSER.parse(parser, null);
+        aggregation.setName(name);
+        return aggregation;
+    }
+
+    public static class ParsedBucket extends ParsedMultiBucketAggregation.ParsedBucket<Double> implements Histogram.Bucket {
+
+        static ParsedBucket fromXContent(XContentParser parser, boolean keyed) throws IOException {
+            final ParsedBucket bucket = new ParsedBucket();
+
+            XContentParser.Token token = parser.currentToken();
+            String currentFieldName = parser.currentName();
+
+            if (keyed) {
+                ensureExpectedToken(XContentParser.Token.FIELD_NAME, token, parser::getTokenLocation);
+                bucket.setKeyedString(currentFieldName);
+                ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
+            }
+
+            List<Aggregation> aggregations = new ArrayList<>();
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                if (token == XContentParser.Token.FIELD_NAME) {
+                    currentFieldName = parser.currentName();
+                } else if (token.isValue()) {
+                    if (CommonFields.KEY_AS_STRING.getPreferredName().equals(currentFieldName)) {
+                        bucket.setKeyAsString(parser.text());
+                    } else if (CommonFields.KEY.getPreferredName().equals(currentFieldName)) {
+                        bucket.setKey(parser.doubleValue());
+                    } else if (CommonFields.DOC_COUNT.getPreferredName().equals(currentFieldName)) {
+                        bucket.setDocCount(parser.longValue());
+                    }
+                } else if (token == XContentParser.Token.START_OBJECT) {
+                    String typeAndName = parser.currentName();
+                    int delimiterPos = typeAndName.indexOf(Aggregation.TYPED_KEYS_DELIMITER);
+                    String type;
+                    String name;
+                    if (delimiterPos > 0) {
+                        type = typeAndName.substring(0, delimiterPos);
+                        name = typeAndName.substring(delimiterPos + 1);
+                        aggregations.add(parser.namedObject(Aggregation.class, type, name));
+                    } else {
+                        throw new ParsingException(parser.getTokenLocation(),
+                                "Cannot parse bucket's aggregation without type information. Set [" + RestSearchAction.TYPED_KEYS_PARAM
+                                        + "] parameter on the request to ensure the type information is added to the response output");
+                    }
```

Comment:
I would try to share this block of code that parses an inner aggregation. It is going to be needed in many places and it should be exactly the same everywhere.